how-to-build-with-ddn docs for SQL Server (#915)
hasura/ddn-docs: docs/how-to-build-with-ddn/with-sqlserver.mdx

1. The [SQL Server extension](https://marketplace.visualstudio.com/items?itemName=ms-mssql.mssql) is available for
   [Visual Studio Code](https://code.visualstudio.com/).

2. [Azure Data Studio](https://azure.microsoft.com/en-us/products/data-studio) for MacOS, Linux, and Windows.

3. The [Go SQLCMD CLI](https://github.com/microsoft/go-sqlcmd) connects to your SQL Server database from the command
   line for Windows, MacOS, and Linux.

4. `mssql-tools` CLI:

   3.1. Install
   [mssql-tools](https://learn.microsoft.com/en-us/sql/linux/sql-server-linux-setup-tools?view=sql-server-ver16&tabs=redhat-install)

   3.2. Run the following in a bash shell:

   ```sh
   docker exec -it sqlserver /opt/mssql-tools18/bin/sqlcmd -S localhost -U SA -P 'Password123' -d 'master' -C
   ```

   3.2.1 The flag `-C` is used to apply the `trust server certificate` set to true. This should only be used for local
   database and not on production databases.

   3.2.2 Each SQL statement in the `mssql-tools` interactive shell needs a `GO` statement at the end for the SQL
   statement to process. For example,

   ```
   SELECT * FROM users;
   GO
   ```

> :::

Reviewer: I recommend we drop this admonition and instead replace it with only the instructions on using `exec` with Docker. 

SQL Server users will know how they like to connect and — if they don't want to reach for what we've laid out — they can use their tool of choice 👍

### Step 8. Start your local services

Use the `docker-start` script in the Hasura context to start your local services.

```sh title="Start your local Hasura DDN Engine and SQL Server connector:"
ddn run docker-start
```

Your terminal will be taken over by logs for the different services.

### Step 9. Run your first query

```sh title="In a new terminal tab, open your local console:"
ddn console --local
```

```graphql title="In the GraphiQL explorer of the console, write this query:"
query {
  users {
    userId
    name
    age
  }
}
```

```json title="You'll get the following response:"
{
  "data": {
    "users": [
      {
        "userId": 1,
        "name": "Alice",
        "age": 25
      },
      {
        "userId": 2,
        "name": "Bob",
        "age": 30
      },
      {
        "userId": 3,
        "name": "Charlie",
        "age": 35
      }
    ]
  }
}
```

### Step 10. Iterate on your SQL Server schema

```sh title="Let's add a new table for posts:"
CREATE TABLE posts (
  user_id int,
  post_id int primary key,
  title varchar(255),
  content varchar(255)
  FOREIGN KEY (user_id) REFERENCES users(user_id)
);

> ```

Reviewer: For this and the subsequent SQL statements, multi-line input fails on Docker exec. Let's modify this to a one-liner like the initial statements.

```sh title="Next, create a new build:"
ddn supergraph build local
```

#### Step 11.5 Restart your services

```sh title="Bring everything back up:"
ddn run docker-start
```

### Step 12. Query your new build

```graphql title="Head back to your console and query the posts model:"
query GetPosts {
  posts {
    userId
    postId
    title
    content
  }
}
```

```json title="You'll get a response like this:"
{
  "data": {
    "posts": [
      {
        "userId": 1,
        "postId": 1,
        "title": "My First Post",
        "content": "This is Alice's first post."
      },
      {
        "userId": 1,
        "postId": 2,
        "title": "Another Post",
        "content": "Alice writes again!"
      },
      {
        "userId": 2,
        "postId": 3,
        "title": "Bob's Post",
        "content": "Bob shares his thoughts."
      },
      {
        "userId": 3,
        "postId": 4,
        "title": "Hello World",
        "content": "Charlie joins the conversation."
      }
    ]
  }
}
```

### Step 13. Create a relationship

```sh title="Since there's already a foreign key on the posts table in SQL Server, we can easily add the relationship:"

> ddn relationship add my_sqlserver "posts"

Reviewer: ```suggestion
ddn relationship add my_sqlserver posts
```

data seeded; Hasura will never modify your source schema.

<Prereqs />

## Tutorial

### Step 1. Authenticate your CLI

```sh title="Before you can create a new Hasura DDN project, you need to authenticate your CLI:"
ddn auth login
```

This will launch a browser window prompting you to log in or sign up for Hasura DDN. After you log in, the CLI will
acknowledge your login, giving you access to Hasura Cloud resources.

### Step 2. Scaffold out a new local project

```sh title="Next, create a new local project:"
ddn supergraph init my-project && cd my-project
```

Once you move into this directory, you'll see your project scaffolded out for you. You can view the structure by either
running `ls` in your terminal, or by opening the directory in your preferred editor.

### Step 3. Initialize your SQL Server connector

```sh title="In your project directory, run:"
ddn connector init my_sqlserver -i

```

From the dropdown, start typing `sqlserver` and hit enter to accept the default port. Then, provide the following
values:

**Connection string**

```plaintext
Server=<hostname>,<port>;Database=master;Uid=sa;Password=Password123;TrustServerCertificate=true
```

:::note

> If the database is created using the local docker setup, it needs to be hosted on cloud using applicarions such as

Reviewer 2: ```suggestion
If the database is running locally, then the database should be made accessible to the internet using tools like [ngrok](https://ngrok.com/downloads/linux)
```

Author: Done

[ngrok](https://ngrok.com/downloads/linux). For example,

```plaintext
Server=0.tcp.in.ngrok.io,16543;Database=master;Uid=sa;Password=Password123;TrustServerCertificate=true
```

```sh title="after the installation, host the database using:"
ngrok tcp 1433
```

:::

:::note

`TrustServerCertificate=true` should only be added for local databases.

:::

### Step 4. Start the local SQL Server container

```sh title="Start a SQL Server instance"
docker pull mcr.microsoft.com/mssql/server:2019-latest
docker run -e "ACCEPT_EULA=Y" -e "MSSQL_SA_PASSWORD=Password123" -p 1433:1433 --name sqlserver -d mcr.microsoft.com/mssql/server:2019-latest
```

Download [Visual Studio Code](https://code.visualstudio.com/) and install the
[SQL Server extension](https://marketplace.visualstudio.com/items?itemName=ms-mssql.mssql) in it.

> 

Reviewer 2: Why do we need this?

```plaintext title="Connect to the database using the connection string:"
Server=localhost,1433;Database=master;Uid=sa;Password=Password123;TrustServerCertificate=true
```

```SQL title="Create a table in the database:"
CREATE TABLE users (user_id int, name varchar(255), age int);
```

```SQL title="Then, seed the table:"
INSERT INTO users (user_id, name, age) VALUES (1, 'Alice', 25), (2, 'Bob', 30), (3, 'Charlie', 35);
```

```SQL title="You can verify this by running:"
SELECT * FROM users;
```

You should see a list of users returned.

:::tip

> Alternativey, other applications and tools can also be used to seed the database, such as

Reviewer 2: ```suggestion
Alternatively, other applications and tools can also be used to seed the database, such as
```

   2. Each SQL statement in the mssql-tools interactive shell needs a `GO` statement at the end for the SQL statement to
      process. For example,
      ```
      SELECT * FROM users;
      GO
      ```

   :::

:::

### Step 5. Introspect your SQL Server database

```sh title="Next, use the CLI to introspect your SQL Server database:"
ddn connector introspect my_sqlserver
```

After running this, you should see a representation of your database's schema in the
`app/connector/my_sqlserver/configuration.json` file; you can view this using `cat` or open the file in your editor.

```sh title="Additionally, you can check which resources are available — and their status — at any point using the CLI:"
ddn connector show-resources my_sqlserver
```

### Step 6. Add your model

```sh title="Now, track the table from your SQL Server database as a model in your DDN metadata:"
ddn models add my_sqlserver users
```

Open the `app/metadata` directory and you'll find a newly-generated file: `Users.hml`. The DDN CLI will use this Hasura
Metadata Language file to represent the `users` table from SQL Server in your API as a
[model](/reference/metadata-reference/models.mdx).

> 

Reviewer 2: Let's also add a comment on how users can track all the available models using `ddn models add *`

CREATE TABLE posts (
  user_id int,
  post_id int,
  title varchar(255),
  content varchar(255)
);
```

```sh title="Then, seed it:"
INSERT INTO posts (user_id, post_id, title, content) VALUES
  (1, 1, 'My First Post', 'This is Alice''s first post.'),
  (1, 2, 'Another Post', 'Alice writes again!'),
  (2, 3, 'Bob''s Post', 'Bob shares his thoughts.'),
  (3, 4, 'Hello World', 'Charlie joins the conversation.');
```

```sh title="Finally, we can check the posts were generated:"
SELECT * FROM posts;
```

### Step 11. Refresh your metadata and rebuild your project

:::tip

The following steps are necessary each time you make changes to your **source** schema. This includes, adding,
modifying, or dropping tables.

:::

#### Step 11.1. Re-introspect your data source

```sh title="Run the introspection command again:"
ddn connector introspect my_sqlserver
```

In `app/connector/my_sqlserver/configuration.json`, you'll see schema updated to include operations for the `posts`
table. In `app/metadata/my_sqlserver.hml`, you'll see `posts` present in the metadata as well.

#### Step 11.2. Update your metadata

```sh title="Add the posts model:"
ddn model add my_sqlserver posts

> ```

Author: We'll have to add syntax to add `foreign keys` first also here in this right?

query GetPosts {
  posts {
    userId
    postId
    title
    content
  }
}
```

```json title="You'll get a response like this:"
{
  "data": {
    "posts": [
      {
        "userId": 1,
        "postId": 1,
        "title": "My First Post",
        "content": "This is Alice's first post."
      },
      {
        "userId": 1,
        "postId": 2,
        "title": "Another Post",
        "content": "Alice writes again!"
      },
      {
        "userId": 2,
        "postId": 3,
        "title": "Bob's Post",
        "content": "Bob shares his thoughts."
      },
      {
        "userId": 3,
        "postId": 4,
        "title": "Hello World",
        "content": "Charlie joins the conversation."
      }
    ]
  }
}
```

### Step 13. Create a relationship

```yaml title="Open the Posts.hml file and add the following to the end:"
---
kind: Relationship
version: v1
definition:
  name: user
  sourceType: Posts
  target:
    model:
      name: Users
      relationshipType: Object
  mapping:
    - source:
        fieldPath:
          - fieldName: userId
      target:
        modelField:
          - fieldName: userId

> ```

Reviewer 2: Since we support native relationships in SQL server, this manual step seems unnecessary and this can be automatically added by `ddn model add "*"`

Author: When I added the model, I didn't see the relationship info. I followed the steps that are mentioned in this doc.

Author: Done
